### DISCUSSION TITLE
Why does `brew shellenv` put linuxbrew in `$PATH` before /usr/bin/ ?

### DISCUSSION THREAD

**Discussion by pmorch** [Homebrew/discussions#2560] [Linux] [answered]
The recommendation is to put:

    eval "$(brew shellenv)"

in one of my dotfiles. When I run `brew shellenv` it puts `/home/linuxbrew/.linuxbrew/bin` in the path before `/usr/bin` which causes all apps installed with `apt` that use `python3` to fail.

Why put `linuxbrew`'s `bin` before `/usr/bin` in the `$PATH`, when it breaks `apt` packages, whereas apps installed with `brew install` work either way?

Running on Ubuntu 20.04 LTS under WSL2.

## Details

When I run `brew shellenv` I get:

    export PATH="/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin${PATH+:$PATH}"; 

So `/home/linuxbrew/.linuxbrew/bin/python3` is before `/usr/bin/python3` in my `$PATH`. This messes up any packages that I've installed with `apt`, like e.g.:

```
$ ansible
Traceback (most recent call last):
  File "/usr/bin/ansible", line 34, in <module>
    from ansible import context
ModuleNotFoundError: No module named 'ansible'
```

(Yes, I could've installed `ansible` with `brew`, but operations wants to use the (old) system ansible and it gets installed much earlier than `brew`.)

`brew` installed `python3` because it is a dependency for `jrnl`. But `jrnl` is careful to use the "correct" python3, whereas ansible is not.

```
$ head -n 1 $(which jrnl)
#!/home/linuxbrew/.linuxbrew/Cellar/jrnl/2.8.3_1/libexec/bin/python3.9

$ head -n 1 $(which ansible)
#!/usr/bin/env python3
```

As far as I can see, everything works if I swap the order in the `$PATH`. But that goes against `brew`'s recommendation and I'd like to understand why `brew shellenv` produces the output it does, and what problems I'll likely run into if I proceed with my non-standard `$PATH`. For me/us, `brew` breaking `apt` packages is rather bad...

**Comment by SMillerDev** [comment, marked as answer]
> Why put linuxbrew's bin before /usr/bin in the $PATH, when it breaks apt packages, whereas apps installed with brew install work either way?

Because brew does not expect you to use a second package manager at the same time (and neither does apt). As for the second part of this, there is no guaranteeing that brew apps work either way. We try our best to make packages work with our dependencies but we don't develop them so we can't be sure it'll work.

**Reply by pmorch** [reply]
Reading this again, I don't understand how this answer makes sense.

> Because brew does not expect you to use a second package manager at the same time (and neither does apt).

`apt` doesn't, and usually there isn't a second package manager in addition to `apt`.

But for `brew`, doesn't virtually every single linux distribution use a package manager? Is it even possible to create a linux system with `brew` but without another package manager?

It seems to me that if `brew` doesn't expect there to be another package manager, that is a bug, because there (almost?) always is.

**Comment by carlocab** [comment]
You can reverse the order of Homebrew's prefix and `/usr/bin` if that's what works for you. The output of `brew shellenv` is just designed for the most common use-case for Homebrew, but that doesn't mean it works for all of them.

**Reply by pmorch** [reply]
I get it that Homebrew comes from MacOS and all. But on Linux, isn't the most sane default to not risk breaking `apt`'s packages?

**Reply by SMillerDev** [reply]
Well, we can't really assume that apt exists. The most common case for homebrew on Linux is "I'm not allowed to install anything as root so I use homebrew". So conflicts are relatively low, because people with a new OS don't have many reasons to use homebrew.

**Reply by carlocab** [reply]
> But on Linux, isn't the most sane default to not risk breaking `apt`'s packages?

No, because one of the most common use cases for Homebrew is to replace tools found in `/usr/bin`. You can't do that if `/usr/bin` comes earlier in your `PATH`.

**Reply by pmorch** [reply]
> No, because one of the most common use cases for Homebrew is to replace tools found in `/usr/bin`

Point taken. I understand now.

**Reply by pmorch** [reply]
And just in case you're interested:

> ... because people with a new OS don't have many reasons to use homebrew.

FYI: In case any body is interested, my use case is, that very many of the `kubernetes` tools don't have `apt` packages, but they do have `brew` packages, so I use `apt` and `brew` for e.g.:

* k9s
* kind
* stern
* yq
* kbld
* jsonnet
* krew

but interesting to note that this is unusual.

**Comment by pmorch** [comment]
So I guess I have no choice but to decide whether I prefer `apt`'s packages or `brew`'s packages to work best... Then I choose `apt`'s.

The best way I could come up with was this for `bash` and `zsh`:

    if [ -z "$HOMEBREW_PREFIX" ]; then
        eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv | egrep -v '\bPATH=')"
        export PATH="$PATH:${HOMEBREW_PREFIX}/bin:${HOMEBREW_PREFIX}/sbin"
    fi

Thank you for your replies, @SMillerDev  and @carlocab

**Reply by gromgit** [reply]
Or you could simply do:
```
eval "$(brew shellenv)"
PATH=/usr/bin:/other/bindir/to/override/Homebrew:$PATH
```
`brew shellenv` is idempotent, so once its `bin` is in `$PATH`, it won't bother adding itself again (and overriding your system `bin` preference in the process).

Every Unix I've ever used only cares about the order in which directories appears. None of them care whether there are duplicates.

**Comment by Merrit** [comment]
FYI, this is still causing breakage in 2024.

[This reply](https://github.com/orgs/Homebrew/discussions/2560#discussioncomment-1716820) by @pmorch seems to still resolve it as a workaround.

**Reply by mabequinho** [reply]
> FYI, this is still causing breakage in 2024.
> 
> [This reply](https://github.com/orgs/Homebrew/discussions/2560#discussioncomment-1716820) by @pmorch seems to still resolve it as a workaround.

this totally breaks Gnome but it should? .bashrc is overriding the entire gui environment, the likes of Geary and GOA. 

EDIT: cleared ~/.bash_profile to prevent sourcing bashrc

**Reply by cache-sk** [reply]
glibc formulae stuck at 2 years old 2.35 is marvellous..

**Reply by SMillerDev** [reply]
That's because that's the minimum requirement. If your OS has a newer version brew can use that.

**Comment by FatalMerlin** [comment]
I just implemented it like this for myself, to avoid duplicates in the path.

I added a check if `brew` is installed because I am using my `.zshrc` on multiple machines.

```bash
# Load homebrew shellenv
BREW=/home/linuxbrew/.linuxbrew/bin/brew
if [ -x "$BREW" -a -z "$HOMEBREW_PREFIX" ]; then
  PREV_PATH=$PATH
  # brew needs tools like dirname and readlink
  PATH=/usr/bin
  # this sets PATH, HOMEBREW_PREFIX, HOMEBREW_CELLAR, HOMEBREW_REPOSITORY
  eval "$($BREW shellenv)"
  export PATH=$PREV_PATH${PATH+:$PATH}
fi
```
